### PR TITLE
Skeleton Trait Implementation

### PR DESCRIPTION
**Pull request by luiscuenca**
https://highfidelity.manuscript.com/f/cases/22182/Implement-Avatar-Traits

This PR creates an Avatar Skeleton Trait that provides enough information to the client to represent the avatar skeleton before loading its fbx file.
A new menu item has been created in order to draw a stick figure based on this skeleton data.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18878/HighFidelity-Beta-PR15438-4a236c3.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18878/HighFidelity-Quest-Beta-PR15438-4a236c3.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18881/HighFidelity-Beta-PR15438-9c90a2b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18881/HighFidelity-Beta-Interface-PR15438-9c90a2b.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18881/HighFidelity-Beta-PR15438-9c90a2b.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18881/HighFidelity-Quest-Beta-PR15438-9c90a2b.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18881/HighFidelity-Beta-PR15438-9c90a2b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18881/HighFidelity-Beta-Interface-PR15438-9c90a2b.exe

**Comment by luiscuenca** [comment]
## TEST PLAN
- Go to a domain with other avatars and activate Developer/Avatar/**Debug Draw Other Skeletons**. Other avatar meshes will not render and instead a stick figure should render.

File: interface/src/avatar/OtherAvatar.cpp [hunk 1]
```diff
@@ -358,6 +359,61 @@ void OtherAvatar::simulate(float deltaTime, bool inView) {
...
+            auto jointTranslation = jointScale * (jointData[i].translationIsDefaultPose ? skeletonData[i].defaultTranslation : jointData[i].translation);
+            if (skeletonData[i].jointName == "Hips") { 
+                jointTranslation = glm::vec3(0.0f);
+            }
+            jointPoses.push_back(AnimPose(jointRotation, jointTranslation));
+            jointNames.push_back(skeletonData[i].jointName);
+        }
+        QVector<AnimPose> worldFramePoses;
+        for (int i = 0; i < (int)jointPoses.size(); i++) {
+            auto &jointPose = jointPoses[i];
```
Comment (Atlante45) [inline interface/src/avatar/OtherAvatar.cpp]: Coding Standard: `&` should be on the type, not the variable.
(There a few I've noticed in the PR, just gonna comment on that one)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18914/HighFidelity-Beta-PR15438-b10aca1.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18914/HighFidelity-Beta-Interface-PR15438-b10aca1.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18914/HighFidelity-Beta-PR15438-b10aca1.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18914/HighFidelity-Quest-Beta-PR15438-b10aca1.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18914/HighFidelity-Beta-PR15438-b10aca1.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18914/HighFidelity-Beta-Interface-PR15438-b10aca1.exe

File: interface/src/avatar/OtherAvatar.cpp [hunk 2]
```diff
@@ -358,6 +359,61 @@ void OtherAvatar::simulate(float deltaTime, bool inView) {
...
+            int parentIndex = skeletonData[i].parentIndex;
+            if (parentIndex < worldFramePoses.size()) {
+                glm::vec3 xAxis = jointPose.rot() * Vectors::UNIT_X;
+                glm::vec3 yAxis = jointPose.rot() * Vectors::UNIT_Y;
+                glm::vec3 zAxis = jointPose.rot() * Vectors::UNIT_Z;
+
+                auto parentRotation = parentIndex > -1 ? jointPoses[parentIndex].rot() : Quaternions::IDENTITY;
+                auto parentPosition = parentIndex > -1 ? worldFramePoses[parentIndex].trans() : rigToAvatar.trans();
+                auto jointPosition = parentPosition + parentRotation * (jointPose.trans() * METERS_PER_CENTIMETER);
+                worldFramePoses.push_back(AnimPose(jointPose.rot(), jointPosition));
```
Comment (hyperlogic) [inline interface/src/avatar/OtherAvatar.cpp]: I don't think this is correct.  The order that poses will be pushed on the worldFramePoses vector will not match the same order as the joints themselves.  So using the parentIndex as an index into worldFramePoses sometimes will not return the joint you expect.

I'll suggest an alternative approach.  In the first loop, `for (int i = 0; i < jointData.size(); i++) {`, transform each jointPose into world space, by mulitplying in the parent joint transform as well.  For an example of this look at AnimSkeleton::convertRelativePosesToAbsolute().  Then you can just draw a line between the parent and the child bone.  I don't even think you need the second loop.

File: libraries/avatars-renderer/src/avatars-renderer/Avatar.cpp [hunk 3]
```diff
@@ -1449,6 +1449,37 @@ QStringList Avatar::getJointNames() const {
     return result;
 }
 
+std::vector<AvatarSkeletonTrait::UnpackedJointData> Avatar::getSkeletonDefaultData() {
+    std::vector<AvatarSkeletonTrait::UnpackedJointData> defaultSkeletonData;
+    if (_skeletonModel->isLoaded()) {
+        auto& model = _skeletonModel->getHFMModel();
+        auto& rig = _skeletonModel->getRig();
+        float geometryToRigScale = glm::length(extractScale(rig.getGeometryToRigTransform()));
```
Comment (hyperlogic) [inline libraries/avatars-renderer/src/avatars-renderer/Avatar.cpp]: Taking the length of the scale doesn't seem correct to me.  In this case you can assume that the scale is uniform and just take one of the parameters.   For example: if the scale is (1.0, 1.0, 1.0) the length of this vector is actually sqrt(3) which isnt what you want.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19000/HighFidelity-Beta-PR15438-080f891.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19000/HighFidelity-Beta-Interface-PR15438-080f891.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19000/HighFidelity-Beta-PR15438-080f891.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19000/HighFidelity-Beta-Interface-PR15438-080f891.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19000/HighFidelity-Beta-PR15438-080f891.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19000/HighFidelity-Quest-Beta-PR15438-080f891.apk)

**Pull request merged**